### PR TITLE
Add assignment 2

### PR DESCRIPTION
**Pull request by adifeiner**
Add BinaryTree, BinaryTreeNode and Main

### CODE REVIEW

File: BinaryTree.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
+public class BinaryTree<T> {
+
+    private BinaryTreeNode<T> root;
```
Comment (brzozan) [inline BinaryTree.java]: Root field seems to never be changed, since there is no remove operation on this tree. In such case it is usually a good idea to make field final, so that compiler can detect unexpected assignments.

File: BinaryTree.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    public BinaryTree(BinaryTreeNode<T> root) {
+        this.root = root;
+    }
+
+    public BinaryTreeNode<T> getRoot() {
+        return root;
+    }
+
+    // Check if the key exists in the tree
+    public Boolean isExist(T key) {
```
Comment (brzozan) [inline BinaryTree.java]: exists() or doesExist() would be better.

File: BinaryTree.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        return isExist(root, key);
+    }
+
+    // Recursively check if the key exists in the tree (receives root)
+    public Boolean isExist(BinaryTreeNode<T> root, T key) {
```
Comment (brzozan) [inline BinaryTree.java]: Ditto.

Also please prefer non-boxed types rather than boxed ones, in this case boolean rather than Boolean.

File: BinaryTree.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
```
Comment (brzozan) [inline BinaryTree.java]: Nitpick: no need for empty lines at the beginning of methods. Both here and in other places.

File: BinaryTree.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+
+    // The function searches the node with the key, if the key exists the function
+    // return it, otherwise, return null
+    public BinaryTreeNode<T> findNode(T key) {
+        if (this.isExist(key)) {
+            return findNode(root, key);
+        }
+        System.out.println("The node doesn't exist");
```
Comment (brzozan) [inline BinaryTree.java]: It is usually better to remove debug statements from code before submission, especially in methods that can potentially be called multiple times, like getters/setters/finders etc.

File: BinaryTree.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        return false;
+    }
+
+
+    // The function searches the node with the key, if the key exists the function
+    // return it, otherwise, return null
+    public BinaryTreeNode<T> findNode(T key) {
+        if (this.isExist(key)) {
+            return findNode(root, key);
```
Comment (brzozan) [inline BinaryTree.java]: This requires the tree to be traversed twice - once to check if node exists, second time to actually find it. Could code be refactored to have only one tree traversal?
Also please look at the code of isExist and findNode - they look very similar...

File: BinaryTreeNode.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
+public class BinaryTreeNode<T> {
+
+        private T data;
```
Comment (brzozan) [inline BinaryTreeNode.java]: Since data is set only in constructor and not changed anywhere, please consider making it final.

File: Main.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
+public class Main {
+
+    private static final boolean left = true;
+    private static final boolean right = false;
```
Comment (brzozan) [inline Main.java]: Typically static final fields are named using capital letters (LEFT, RIGHT), to make them easily distinguishable from regular ones.

You could also consider using enum rather than boolean to drive which side insertion should be done.

File: BinaryTree.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        return foundNode;
+    }
+
+    // The function create a left (left or right) to the node with the value parent
+    public void insert(T value, T parent, boolean toLeft) {
+
+        // Check if this value already exists in the tree
+        if (this.isExist(value)) {
+            System.out.println("This value already exists");
```
Comment (brzozan) [inline BinaryTree.java]: Error/warning messages not intended to end user but rather to developer are typically written to stderr, i.e. System.err.println.

File: Main.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        myTree.insert(19, 18, right);
+        printAncestors(myTree, 5);  // 16, 9, 3
+        System.out.println();
+        System.out.println("Lowest common ancestor: " +
+                lowestCommonAncestor(myTree, myTree.findNode(14), myTree.findNode(5)).getData()); //9
+    }
+
+    // The function receives a Binary Tree and a key
+    // Prints all the ancestors of the key from the root
+    public static <T> void printAncestors(BinaryTree<T> tree, T key) {
```
Comment (brzozan) [inline Main.java]: List of ancestors of a key seems like a potentially useful thing. What if in the future we want to do something interesting with it rather than just print it?
Could you please refactor code to separate finding ancestors from printing them?

Also printAncestors operates on internal data of BinaryTree class. It would be nice to keep operations that touch internal data within the class that contains this data. Could you move methods that operate on BinaryTree to BinaryTree class?

File: Main.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+            if (tree.isExist(root.getLeftChild(), key)) {
+                printAncestors(new BinaryTree(root.getLeftChild()), key);
+            }
+        }
+    }
+
+    // The function receives a Binary Tree and two nodes
+    // Return the lowest common ancestor
+    public static <T> BinaryTreeNode<T> lowestCommonAncestor(BinaryTree<T> tree, BinaryTreeNode<T> firstNode,
```
Comment (brzozan) [inline Main.java]: It feels like this method also belongs to BinaryTree...

And optional thing to think about: could finding lowest common ancestor be simplified by using some code that has been created for printing ancestors of a single node?

File: Main.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        myTree.insert(9, 16 , left);
+        myTree.insert(18, 16, right);
+        myTree.insert(3, 9, left);
+        myTree.insert(14, 9, right);
+        myTree.insert(1, 3, left);
+        myTree.insert(5, 3, right);
+        myTree.insert(19, 18, right);
+        printAncestors(myTree, 5);  // 16, 9, 3
+        System.out.println();
+        System.out.println("Lowest common ancestor: " +
```
Comment (brzozan) [inline Main.java]: Finding LCA has quite a few edge cases (key not present, one key being ancestor of the other etc), while this println tests only a single one. Additionally it requires a person to look at the output and verify if result is correct. This makes it easy for bugs to slip through.

Could you please add some tests that would check different scenarios, verify results returned by tested methods and output status of the test (passed/failed)? You could either use [JUnit](https://github.com/junit-team/junit4/wiki/Getting-started) or go with your own simple methods.

**Review by brzozan — CHANGES_REQUESTED** [review]
Very good. I have added a few, mostly cosmetic and testing-related comments.

File: BinaryTree.java [hunk 13]
```diff
@@ -1,6 +1,15 @@
+import java.util.ArrayList;
+import java.util.List;
+
 public class BinaryTree<T> {
+    private final BinaryTreeNode<T> root;
+    public enum InsertChild {Left, Right};
+
+    private InsertChild m_insertChild = InsertChild.Left;
```
Comment (brzozan) [inline BinaryTree.java]: Is storing direction of child insertion inside binary tree really needed? Having container in a state that would only insert to the left or only to the right does not seem too useful and would be a pain to use.

Why not just passing this enum as parameter to insert method?

File: BinaryTree.java [hunk 14]
```diff
@@ -15,99 +24,133 @@ public BinaryTree(BinaryTreeNode<T> root) {
...
+    public BinaryTreeNode exists(T key) {
+        return exists(root, key);
     }
 
     // Recursively check if the key exists in the tree (receives root)
-    public Boolean isExist(BinaryTreeNode<T> root, T key) {
-
-        if (root == null){
-            return false;
+    public BinaryTreeNode exists(BinaryTreeNode<T> root, T key) {
```
Comment (brzozan) [inline BinaryTree.java]: Why not make `exists` a method in BinaryTreeNode, so that it can operate on `this` instead of explicitly passed root?
And again, containsKey would be a better name.

File: BinaryTree.java [hunk 15]
```diff
@@ -15,99 +24,133 @@ public BinaryTree(BinaryTreeNode<T> root) {
     }
 
     // Check if the key exists in the tree
-    public Boolean isExist(T key) {
-        return isExist(root, key);
+    public BinaryTreeNode exists(T key) {
```
Comment (brzozan) [inline BinaryTree.java]: In this form it should be something like `findNode`, since it returns something rather than boolean flag that would say if something exists.
I think even better name would be something like `containsKey`, returning boolean. `tree.exists()` feels more like checking if the tree exists rather than whether it contains given key.

File: BinaryTree.java [hunk 16]
```diff
@@ -15,99 +24,133 @@ public BinaryTree(BinaryTreeNode<T> root) {
...
+                System.err.println("There is already right child");
                 return;
             }
         }
     }
-}
+
+    // The function receives a Binary Tree and a key
+    // Prints all the ancestors of the key from the root
+    public List<T> printAncestors(BinaryTree<T> tree, T key) {
```
Comment (brzozan) [inline BinaryTree.java]: This method is doing two things - prints and returns list. This is considered bad practice, as user of the API cannot get one effect without the other. What if I don't want to print the result but only receive list of ancestors?

It would be much better to have findAncestors (or getAncestors or something similar) that would just return list without printing, and a helper method that would take list as parameter and print it out.

Second thing: is the tree parameter really needed? You already have a tree object - `this`.

File: BinaryTree.java [hunk 17]
```diff
@@ -15,99 +24,133 @@ public BinaryTree(BinaryTreeNode<T> root) {
...
+        BinaryTreeNode<T> root = tree.getRoot();
+
+        if (root == null || firstNode == null || secondNode == null) {
+            System.err.println("Invalid input");
+            return null;
+        }
+
+        if (firstNode.equals(secondNode)) {
+            System.err.println("Both nodes are identical, invalid input");
+            System.err.println("Both nodes are identical, invalid input");
```
Comment (brzozan) [inline BinaryTree.java]: Duplicate message.

File: BinaryTree.java [hunk 18]
```diff
@@ -15,99 +24,133 @@ public BinaryTree(BinaryTreeNode<T> root) {
...
+            if (exists(root.getLeftChild(), key) != null) {
+                printAncestors(ancestors, root.getLeftChild(), key);
+            }
+        }
+        return ancestors;
+    }
+
+    // The function receives a Binary Tree and two nodes
+    // Return the lowest common ancestor
+    public BinaryTreeNode<T> lowestCommonAncestor(BinaryTree<T> tree, BinaryTreeNode<T> firstNode,
```
Comment (brzozan) [inline BinaryTree.java]: Again, is the tree parameter really needed?

Second thing: you require client of this class to provide instances of `BinaryTreeNode`. This requires user to do some extra things to achieve their goal, which is getting lowestCommonAncestor... Does user of the class really need to care about nodes?
Imagine that you are a user of this class and all you want is to find common ancestor of some keys. What would be the simplest and most convenient API that would let you do it?
What parameters would be most convenient for you (as a user) to pass? What form of result would be most convenient to get in return? How would you like to be informed about failure?

File: BinaryTreeTest.java [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+        assertEquals(result , myTree.printAncestors(myTree, 5));
+
+        assertNull(myTree.printAncestors(myTree, 4));
+
+        result.remove(2);
+        result.remove(1);
+        result.remove(0);
+
+        assertEquals(result, myTree.printAncestors(myTree, 16));
```
Comment (brzozan) [inline BinaryTreeTest.java]: Typically each test method should verify only one scenario and have name that is informative about the case (and expected result if possible - it not always makes sense though). Could you split test methods to single-scenario ones?
For example testPrintAncestors_existingNode and testPrintAncestors_missingNode or something similar.

The same applies to tests for lowestCommonAncestor.

**Review by brzozan — CHANGES_REQUESTED** [review]
Thanks for applying changes. I have some more comments, but we are almost done :)

File: BinaryTree.java (hunk 18 above)
Comment (adifeiner) [reply BinaryTree.java]: I might have misunderstood the question. Does the function need to get two keys and return the key of the lower common ancestor?

How would you like to be informed about failure? - As a user or a developer?

Comment (brzozan) [reply BinaryTree.java]: From the perspective of the code using this class it is much more convenient to just use keys as parameters to the method. Nodes are internal implementation detail that client code should not be concerned with.

Regarding tree parameter, it is not needed, because you already do have tree - it is `this`.

As for notifying err, there are multiple options - using `Optional<T>`, returning null, throwing exception...

File: BinaryTreeTest.java [hunk 20]
```diff
@@ -9,63 +15,71 @@
  * Created by adi on 6/1/2017.
  */
 public class BinaryTreeTest {
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+    private BinaryTree<Integer> myTree;
+    @Before
+    public void setUp(){
+        System.setOut(new PrintStream(outContent));
```
Comment (brzozan) [inline BinaryTreeTest.java]: I would advise against testing output of the class unless there are strict requirements about output format. Why not simplify things and just test whether collection returned from findAncestors contains the right elements?

Errors can safely be tested by just checking for null (which you already do in some tests). If you want to be extra safe, you could use exceptions with a separate class (or one class with different error codes/messages) to signal them and check if correct exception is thrown.

**Review by brzozan — CHANGES_REQUESTED** [review]
Thank you! A few more comments.

File: BinaryTreeTest.java [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+        myTree.insert(3, 9, BinaryTree.InsertChild.LEFT);
+        myTree.insert(1, 3, BinaryTree.InsertChild.LEFT);
+        myTree.insert(18, 16, BinaryTree.InsertChild.RIGHT);
+        myTree.insert(14, 9, BinaryTree.InsertChild.RIGHT);
+        myTree.insert(5, 3, BinaryTree.InsertChild.RIGHT);
+        myTree.insert(19, 18, BinaryTree.InsertChild.RIGHT);
+    }
+
+    @After
+    public void cleanUpStream() {
```
Comment (brzozan) [inline BinaryTreeTest.java]: Not needed any more.

File: BinaryTree.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+        // Check if the parent exists in the tree
+        if (!this.root.containsKey(parent)) {
+            System.err.println("This parent isn't exist in the tree");
+            return;
+        }
+        insert(value, this.root.findNode(parent), childSide);
+    }
+
+    // The function create a left (left or right) to the parent node
+    public void insert(T value, BinaryTreeNode<T> parent, InsertChild childSidet) {
```
Comment (brzozan) [inline BinaryTree.java]: Should probably be private. Only minimal set of methods designed for use from outside should be public. All the methods that expose implementation details (BinaryTreeNodes etc) should be private.

File: BinaryTree.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+                parent.setRightChild(value);
+            } else {
+                System.err.println("There is already right child");
+                return;
+            }
+        }
+    }
+
+    // Prints all the ancestors of the key from the root
+    public void printAncestors(T key) {
```
Comment (brzozan) [inline BinaryTree.java]: Not really needed. If client code needs to print results they can just find ancestors and print them themselves. Always try to keep your code as focused on single thing as possible, in this case `BinaryTree` should focus only on operation on binary tree.

File: BinaryTree.java [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+            System.err.println("There is no such key in the tree");
+            return null;
+        }
+        List<T> ancestors = new ArrayList<T>();
+        ancestors = findAncestors(ancestors, this.root, key);
+
+        return ancestors;
+    }
+
+    public List<T> findAncestors (List<T> ancestors, BinaryTreeNode<T> node, T key) {
```
Comment (brzozan) [inline BinaryTree.java]: private.

File: BinaryTreeTest.java [hunk 25]
```diff
@@ -0,0 +1,77 @@
+import jdk.internal.cmm.SystemResourcePressureImpl;
```
Comment (brzozan) [inline BinaryTreeTest.java]: You really do not want to import such internal things in your code :)

**Review by brzozan — APPROVED** [review]
Hi Adi,

Thanks for applying the changes. The code looks good and I really like your tests.

I approve your PR, but I have a few more minor comments, so please fix them before merging.

**Comment by adifeiner** [comment]
Updating with the fix code, Thank!
